File: 贪心算法/middle/376.摆动序列.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <algorithm>
#include <unordered_set>
#include <unordered_map>
#include <math.h>

using namespace std;
/*
 * @lc app=leetcode.cn id=376 lang=cpp
 *
 * [376] 摆动序列
 */

// @lc code=start
class Solution
{
public:
    int wiggleMaxLength(vector<int> &nums)
    {
        int result = 1;
        int preDiff = 0;
        int curDiff = 0;
        for (int i = 1; i < nums.size(); i++)
        {
            // i-1这个元素和下一个元素i的差
            curDiff = nums[i] - nums[i - 1];
            // 如果中间的数 - 上一个数大于0 同时 下一个数 - 中间的数小于0，中间的数在波峰，满足条件
            // 中间的数 - 上一个数小于于0 同时 下一个数 - 中间的数大于0，中间的数在波谷，满足条件
            if ((curDiff > 0 && preDiff <= 0) || (curDiff < 0 && preDiff >= 0))
            {
                result++;
                preDiff = curDiff;
            }
        }

        return result;
    }
};
// @lc code=end

File: 贪心算法/middle/376.摆动序列.ts
/*
 * @lc app=leetcode.cn id=376 lang=typescript
 *
 * [376] 摆动序列
 */
function wiggleMaxLength(nums: number[]): number {
  let result = 1
  // 少于两个元素，不用判断，直接返回数组长度
  if (nums.length === 1) {
    return result
  }

  // 前一个差值
  let perDiff = 0
  // 当前差值
  let curDiff = 0

  // 数组从第三个数还是判断
  for (let i = 0; i < nums.length - 1; i++) {
    curDiff = nums[i + 1] - nums[i]
    // 这里curDiff 等于0的话当前树和下一个数相等
    // 直接跳到下一个数，不进行任何操作
    if (curDiff === 0) {
      continue
    }

    // 前面大于0，后面小于0，满足条件
    if (perDiff >= 0 && curDiff < 0) {
      perDiff = curDiff
      result++
      continue
    }

    // 前面小于于0，后面大于0，满足条件
    if (perDiff <= 0 && curDiff > 0) {
      perDiff = curDiff
      result++
    }
  }

  return result
};
// @lc code=end

